src/azul/openapi/validation.py

```diff
@@ -0,0 +1,147 @@
+import json
+import logging
+from typing import (
+    Mapping,
+)
+
+from chalice import (
+    BadRequestError,
+)
+import jsonschema
+from more_itertools import (
+    one,
+)
+
+from azul.openapi.schema import (
+    type_lookup,
+)
+from azul.uuids import (
+    InvalidUUIDError,
+    InvalidUUIDVersionError,
+    validate_uuid,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SpecValidator:
+
+    def __init__(self,
+                 content: Mapping = None,
+                 pattern: str = None,
+                 name: str = None,
+                 required: bool = None,
+                 schema: Mapping = None,
+                 **kwargs):
+        """
+        :param kwargs: Used to extract the `in` parameter.
+        """
+        self.content = content
+        self.pattern = pattern
+        self.name = name
+        self.required = required
+        self.schema = schema
+        if self.schema is not None and self.content is not None:
+            raise TypeError('OpenAPI spec enforces that schema and content are mutually exclusive')
+        if self.schema:
+            parameter_schema = self.schema
+        else:
+            for k, v in self.content.items():
+                parameter_schema = v['schema']
+        self.validator = type_lookup(parameter_schema['type'])
+        self.json_validators = dict(jsonschema.Draft7Validator.VALIDATORS)
+        self._in = kwargs['in']
+
+    def validate(self, request_parameter: Mapping[str, str]):
+        """
+        Validates if a given request parameter meets the spec that is
+        specified for the endpoint.
+        :param request_parameter: A dict with the keys 'name' and 'value'.
+        """
+
+        param_name = request_parameter['name']
+        param_value = request_parameter['value']
+
+        if param_name == 'filters':
+            self.schema = self.content['application/json']['schema']
+            try:
+                filter_json = json.loads(param_value)
+            except json.decoder.JSONDecodeError:
+                raise BadRequestError(f'The {param_name!r} parameter is not valid JSON')
+            if type(filter_json) is not dict:
+                raise BadRequestError(f'The {param_name!r} parameter must be a JSON object')
+            else:
+                for facet, filter_ in filter_json.items():
+                    try:
+                        relation, value = one(filter_.items())
+                    except (ValueError, AttributeError):
+                        raise BadRequestError(
+                            f"The 'filters' parameter entry for {filter_!r} must be a single-item"
+                            f" dictionary")
+                    try:
+                        jsonschema.validate({facet: filter_}, self.schema)
+                    except (
+                        jsonschema.exceptions.SchemaError,
```

Comment:
This is a major bug. `SchemaError` is only raised when "an invalid schema itself is encountered". If any of our schemas are invalid you are reporting them as 4XX errors.

If you are seeing `SchemaError`s then there's either a bug in the schemas or your conversion code. Either way, I don't think this is an error we should catch.

You *should* add a test that checks each schema and asserts that a `SchemaError` is *not* raised.

Check elsewhere for this same bug.

Reply:
Added check_schema to ServiceSpecValidation test

Reply:
the schema is validated during the IT, moved it to be a unit-test for easier testing.